lmdeploy/pytorch_poc/models/falcon.py

```diff
@@ -0,0 +1,361 @@
+            emb = emb.float()
+
+        self.cos_cached = emb.cos()[None, :, :]
+        self.sin_cached = emb.sin()[None, :, :]
+
+        self.cos_cached = self.cos_cached.type(dtype)
+        self.sin_cached = self.sin_cached.type(dtype)
+
+    def patched_cos_sin(self,
+                        position_ids: torch.Tensor,
+                        device='cpu',
+                        dtype=torch.bfloat16) -> torch.Tensor:
+        total_length = int(position_ids.max().item()) + 1
+
+        if (self.seq_len_cached is None) or (total_length >
+                                             self.seq_len_cached):
+            self._patched_set_cos_sin_cache(total_length, device, dtype)
+        # position_ids.shape == [1, packed_seq_len]
+        return (
+            self.cos_cached[:, position_ids[0], None, :],
+            self.sin_cached[:, position_ids[0], None, :],
+        )
+
+    def _contiguous_batching_forward(self, query: torch.Tensor,
+                                     key: torch.Tensor,
+                                     position_ids: torch.Tensor):
+        # batch, seq_len, *_ = query.shape
+        cos, sin = self.patched_cos_sin(position_ids,
+                                        device=query.device,
+                                        dtype=query.dtype)
+        return (
+            (query * cos) + (rotate_half(query) * sin),
+            (key * cos) + (rotate_half(key) * sin),
+        )
+
+    def forward(self, query, key, position_ids_or_past_key_values_length=0):
+        use_origin = False
+
+        if use_origin:
+            return self.origin_mod(query, key,
+                                   position_ids_or_past_key_values_length)
+        else:
+            return self._contiguous_batching_forward(
+                query, key, position_ids_or_past_key_values_length)
+
+
+class PatchedFalconAttention(nn.Module):
+
+    def _contiguous_batching_forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: torch.Tensor,
+        alibi: Optional[torch.Tensor],
+        attention_mask: torch.Tensor,
+        layer_past: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        use_cache: bool = False,
+        output_attentions: bool = False,
+    ):
+        # prepare inputs for continuous batch forwarding
+        context = self.context.context
+        history_lengths = context.history_lengths
+        q_start_loc, q_seq_length = self.context.q_seq_info
+        history_lengths = q_seq_length.new_tensor(history_lengths)
+        kv_seq_length = q_seq_length + history_lengths
+        max_seq_len = q_seq_length.max().item()
+
+        fused_qkv = self.query_key_value(
+            hidden_states)  # [batch_size, seq_length, 3 x hidden_size]
+
+        # 3 x [batch_size, seq_length, num_heads, head_dim]
+        (query_layer, key_layer, value_layer) = self._split_heads(fused_qkv)
```

Comment:
_split_heads and the reshape before the last linear might have the wrong shape when tensor parallelism.

Reply:
We can merge this first and I can support TP in another PR.